From 3e344117bcb43f77e71a16606f7a8b1d29799e01 Mon Sep 17 00:00:00 2001
From: Andrei Cherechesu <andrei.cherechesu@nxp.com>
Date: Mon, 11 Oct 2021 16:46:51 +0300
Subject: [PATCH 24/24] s32g: bl2: Inject 'atf' reserved-memory node

To avoid relying on CONFIG_PRAM mechanism in U-Boot to hide
the top of the first DRAM bank from Linux, BL2 now fixes-up
the U-Boot FDT by adding a 'atf' subnode in the
'/reserved-memory' node.

Issue: ALB-7902
Upstream-Status: Pending 

Signed-off-by: Andrei Cherechesu <andrei.cherechesu@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 plat/nxp/s32g/s32g_bl2_el3.c | 29 +++++++++++++++++++++++++----
 plat/nxp/s32g/s32g_common.mk |  1 +
 2 files changed, 26 insertions(+), 4 deletions(-)

diff --git a/plat/nxp/s32g/s32g_bl2_el3.c b/plat/nxp/s32g/s32g_bl2_el3.c
index f48afa4d3..22826bfac 100644
--- a/plat/nxp/s32g/s32g_bl2_el3.c
+++ b/plat/nxp/s32g/s32g_bl2_el3.c
@@ -10,6 +10,7 @@
 #include <common/bl_common.h>
 #include <common/desc_image_load.h>
 #include <common/debug.h>
+#include <common/fdt_fixup.h>
 #include <drivers/console.h>
 #include <lib/mmio.h>
 #include <lib/optee_utils.h>
@@ -322,6 +323,22 @@ static int ft_fixup_ddr_errata(void *blob)
 }
 #endif
 
+static int ft_fixup_resmem_node(void *blob)
+{
+	int ret;
+	char nodename[21];
+
+	snprintf(nodename, sizeof(nodename), "atf@%x", BL31_BASE);
+
+	ret = fdt_add_reserved_memory(blob, nodename, BL31_BASE, BL31_SIZE);
+	if (ret) {
+		ERROR("Failed to add 'atf' /reserved-memory node");
+		return ret;
+	}
+
+	return 0;
+}
+
 static int ft_fixups(void *blob)
 {
 	size_t size = fdt_totalsize(blob);
@@ -336,8 +353,12 @@ static int ft_fixups(void *blob)
 
 #if (ERRATA_S32G2_050543 == 1)
 	ret = ft_fixup_ddr_errata(blob);
+	if (ret)
+		goto out;
 #endif
 
+	ret = ft_fixup_resmem_node(blob);
+
 out:
 	flush_dcache_range((uintptr_t)blob, size);
 	return ret;
@@ -567,10 +588,10 @@ void s32g_el3_mmu_fixup(void)
 	int i;
 
 	/* Check the BL31/BL32/BL33 memory ranges for overlapping */
-	_Static_assert(S32G_BL32_BASE + S32G_BL32_SIZE <= BL33_BASE,
-				"BL32 and BL33 memory ranges overlap!");
-	_Static_assert(BL33_BASE + S32G_BL33_IMAGE_SIZE <= BL31_BASE,
-				"BL33 and BL31 memory ranges overlap!");
+	_Static_assert(S32G_BL32_BASE + S32G_BL32_SIZE <= BL31_BASE,
+				"BL32 and BL31 memory ranges overlap!");
+	_Static_assert(BL31_BASE + BL31_SIZE <= BL33_BASE,
+				"BL31 and BL33 memory ranges overlap!");
 
 	/* The calls to mmap_add_region() consume mmap regions,
 	 * so they must be counted in the static asserts
diff --git a/plat/nxp/s32g/s32g_common.mk b/plat/nxp/s32g/s32g_common.mk
index 6e364b2dc..9c944989d 100644
--- a/plat/nxp/s32g/s32g_common.mk
+++ b/plat/nxp/s32g/s32g_common.mk
@@ -86,6 +86,7 @@ BL2_SOURCES		+= plat/nxp/s32g/s32g_lowlevel_bl2.S \
 			   plat/nxp/s32g/s32g_storage.c \
 			   drivers/io/io_storage.c \
 			   common/desc_image_load.c \
+			   common/fdt_fixup.c \
 			   drivers/mmc/mmc.c \
 			   drivers/nxp/s32g/io/io_mmc.c \
 			   drivers/nxp/s32g/io/io_memmap.c \
-- 
2.17.1

